### DISCUSSION TITLE
How to enable a command based on a valid view model?

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#15541] [Q&A]
My view models implement `INotifyDataErrorInfo` and may report custom errors as well as annotate properties with attributes that derive from `ValidationAttribute`.
My question is what is the simplest way to bind a command's CanExecute state to whether the view model as a whole is valid?

**Comment by workgroupengineering** [comment]
```c#
[DependsOn(nameof(HasErrors))]
bool CanSave(object? parameter) =>
  !HasErrors;

public bool HasErrors
{
   get => _HasErrors;
   set => SetProperty( ref _ HasErrors, value);
}
```

**Reply by AArnott** [reply]
That... doesn't actually answer the question. I know how to define a property. The magic I'm looking for is in what _sets_ it.

**Comment by stevemonaco** [comment]
The INDEI interface has a `HasErrors` property. There's also an `ErrorsChanged` event. Implementation is going to heavily rely on your MVVM framework: Mvvm Toolkit, ReactiveUI, custom, or other. ie. Whether you can push (Rx) or pull this state or need to manually raise `ICommand.CanExecuteChanged`.

**Reply by thevortexcloud** [reply]
Similarily you can also just implement a custom `ICommand` that subscribes to the ErrorsChanged Event. There are a lot of ways to solve this. I personally normally just use a ReactiveCommand that refreshes whenever any of the properties that have validation change, then use that to check if the validation has passed and anything else I need to check.

**Reply by AArnott** [reply]
The `INotifyDataErrorInfo.HasErrors` property is not good enough. First off, _I_ implement it, and I'm looking for a simpler way that doesn't require that. Secondly, I've learned that `INotifyDataErrorInfo` should only report errors that are not already captured by validation attributes in order to avoid the UI displaying double error messages. So the interface won't tell the whole story. 
Yet, I know _something_ in Avalonia already has a comprehensive idea as to the validity of the view model's components because errors are displayed in the UI whether they come from property validation attributes or the `INotifyDataErrorInfo` interface.
What I'm looking for is some API in avalonia that just produces an `IObservable<bool>` indicating whether there are no errors across the whole view model. 
I eventually created one myself, which aggregates my own accumulation of results from executing the validation attributes with the `INotifyDataErrorInfo` custom errors. But that's a lot of code to have to write for something that is so common it might be used for every "OK" button in any dialog.

**Reply by thevortexcloud** [reply]
>  First off, I implement it, and I'm looking for a simpler way that doesn't require that

If you are using ReactiveUI it has built in support for data validation and works great with Avalonia. You just inherit from the class and then register validation handlers and it does the rest.

https://github.com/reactiveui/ReactiveUI.Validation#inotifydataerrorinfo-support

>, I've learned that INotifyDataErrorInfo should only report errors that are not already captured by validation attributes in order to avoid the UI displaying double error messages. So the interface won't tell the whole story.

You can disable validation plugins for this exact reason.

https://docs.avaloniaui.net/docs/guides/development-guides/data-validation#manage-validationplugins


> Yet, I know something in Avalonia already has a comprehensive idea as to the validity of the view model's components because errors are displayed in the UI whether they come from property validation attributes or the INotifyDataErrorInfo interface.
What I'm looking for is some API in avalonia that just produces an IObservable<bool> indicating whether there are no errors across the whole view model.

Avalonia itself handles this using an abstraction layer of "plugins" that implement this interface.
https://github.com/AvaloniaUI/Avalonia/blob/76a1c70758884abcebc9e3665aa08290a841efa4/src/Avalonia.Base/Data/Core/Plugins/IDataValidationPlugin.cs#L5-L31

I have not tried it but you could probably implement your own custom validation plugin that uses some other interface than `INotifyDataErrorInfo`.

https://docs.avaloniaui.net/docs/guides/development-guides/data-validation#manage-validationplugins

> But that's a lot of code to have to write for something that is so common it might be used for every "OK" button in any dialog.

Typically you would put repetitive code like that into a helper class or the base class for your view model. I typically just made a `ValidatableViewModel` class when I need to implement `INotifyDataErrorInfo` that does it all for me.

**Reply by AArnott** [reply]
Thank you, @thevortexcloud. I use ReactiveUI, and my view models all derive from a base class I declare, which itself derives from `ReactiveObject`. I'll take a look at `ReactiveValidationObject`.

However that doesn't appear to address the core of my question, so let me reduce it to a simpler one:

Assuming I only use attributes for validation:

```cs
class MyViewModel
{
   [Required]
   public string? Name { get; set; }
}
```

How can I then add an `ICommand` (e.g. `ReactiveCommand`) to my view model that is enabled only when all attributed properties are valid?

Avalonia already has code that reflects over the properties and validates them, in order to present error messages next to the textbox control that I see for this simple sample. Is there an API by which Avalonia exposes a bool that indicates whether it found any such errors? 
I get that Avalonia has a plugin model, and that's fine. I like both plugins (the attributes and the `INotifyDataErrorInfo` interface handler). But how do I tap into the overall state, or even just the state from the attribute-based validation plugin?

I'd like to do something like this:

```cs
class MyViewModel
{
   public MyViewModel()
   {
      this.SubmitCommand = ReactiveCommand.Create(this.Submit, this.IsValid); // IsValid is an IObservable<bool>
   }

   public ICommand SubmitCommand { get; }

   [Required]
   public string? Name { get; set; }

   public void Submit() { }
}
```

This seems like such a common requirement that I'm surprised if no such property or helper method already exists to just produce a bool that can do this. Lacking such an API, this is what I've come up with in my own view model base class:

```cs
	private readonly HashSet<string> propertiesWithAnnotationErrors = new(StringComparer.Ordinal);

	public ViewModelBase()
	{
		this.IsValid = this.WhenAnyValue(vm => vm.HasAnyErrors, hasErrors => !hasErrors);
		this.PropertyChanged += (sender, e) =>
		{
			if (e.PropertyName is null or nameof(this.HasAnyErrors))
			{
				return;
			}

			if (e.PropertyName is nameof(this.HasErrors))
			{
				this.RaisePropertyChanged(nameof(this.HasAnyErrors));
				return;
			}

			if (this.GetType().GetProperty(e.PropertyName, BindingFlags.Instance | BindingFlags.Public) is PropertyInfo property)
			{
				bool oldHasAnyErrorsValue = this.HasAnyErrors;
				if (this.CheckPropertyAnnotations(property) && oldHasAnyErrorsValue != this.HasAnyErrors)
				{
					this.RaisePropertyChanged(nameof(this.HasAnyErrors));
				}
			}
		};

		// Initialize default validity state.
		foreach (PropertyInfo property in this.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public))
		{
			this.CheckPropertyAnnotations(property);
		}
	}

	/// <summary>
	/// Gets a value indicating whether this view model has any errors, including from validation attributes
	/// on its members.
	/// </summary>
	protected bool HasAnyErrors => this.propertiesWithAnnotationErrors.Count > 0;

	/// <summary>
	/// Gets an observable value indicating whether everything in this view model is valid.
	/// </summary>
	/// <remarks>
	/// This is always the inverse of <see cref="HasAnyErrors"/>.
	/// </remarks>
	protected IObservable<bool> IsValid { get; }

	private bool CheckPropertyAnnotations(PropertyInfo property)
	{
		ValidationContext context = new(this)
		{
			DisplayName = property.GetCustomAttribute<DisplayNameAttribute>()?.DisplayName ?? property.Name,
			MemberName = property.Name,
		};

		bool hasErrors = property.GetCustomAttributes<ValidationAttribute>(inherit: true)
			.Any(att => att.GetValidationResult(property.GetValue(this), context) is { ErrorMessage: not null });
		return hasErrors
			? this.propertiesWithAnnotationErrors.Add(property.Name)
			: this.propertiesWithAnnotationErrors.Remove(property.Name);
	}
```

So now any view model derived class can use `this.IsValid` as a way to enable/disable commands when the state is valid.
My actual class also implements and mixes in the `INotifyDataErrorInfo` interface support, but I omitted that in the above snippet for simplicity.

**Reply by jp2masa** [reply]
Why not use `this.WhenAnyValue(x => x.HasErrors).Select(x => !x)` as the `CanExecute` observable?

**Reply by AArnott** [reply]
Because `HasErrors` isn't declared on a typical view model class that has attributed properties. And when it _does_ have that property (because _I_ implemented the `INotifyDataErrorInfo` interface), then I can already do that, because I've written the 100 lines or so to make it happen. But my question is whether (or why not) such an API already exists in Avalonia that can do this for me.

**Reply by jp2masa** [reply]
`ReactiveValidationObject` implements `INotifyDataErrorInfo`, so it should just work.
